core/src/main/java/org/apache/calcite/rel/metadata/RelMdLowerBoundCost.java

```diff
@@ -0,0 +1,127 @@
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.rel.metadata;
+
+import org.apache.calcite.plan.Convention;
+import org.apache.calcite.plan.ConventionTraitDef;
+import org.apache.calcite.plan.RelOptCost;
+import org.apache.calcite.plan.RelOptPlanner;
+import org.apache.calcite.plan.volcano.AbstractConverter;
+import org.apache.calcite.plan.volcano.CascadeRelSet;
+import org.apache.calcite.plan.volcano.CascadeRelSubset;
+import org.apache.calcite.plan.volcano.RelSubset;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.metadata.BuiltInMetadata.LowerBoundCost;
+import org.apache.calcite.util.BuiltInMethod;
+
+/**
+ * Default implementations of the
+ * {@link BuiltInMetadata.LowerBoundCost}
+ * metadata provider for the standard algebra.
+ */
+public class RelMdLowerBoundCost implements MetadataHandler<LowerBoundCost> {
+
+  public static final RelMetadataProvider SOURCE =
+      ReflectiveRelMetadataProvider.reflectiveSource(
+          new RelMdLowerBoundCost(), BuiltInMethod.LOWER_BOUND_COST.method);
+
+  //~ Constructors -----------------------------------------------------------
+
+  protected RelMdLowerBoundCost() {}
+
+  //~ Methods ----------------------------------------------------------------
+
+  public MetadataDef<LowerBoundCost> getDef() {
+    return BuiltInMetadata.LowerBoundCost.DEF;
+  }
+
+  private boolean isLogical(RelNode relNode) {
+    return relNode.getTraitSet().getTrait(ConventionTraitDef.INSTANCE)
+        == Convention.NONE;
+  }
+
+  public RelOptCost getLowerBoundCost(CascadeRelSubset subset,
+      RelMetadataQuery mq, RelOptPlanner planner) {
+
+    if (isLogical(subset)) {
+      // currently only support physical, will improve in the future
+      return null;
+    }
+
+    RelOptCost winner = subset.getWinnerCost();
+    if (winner != null) {
+      // when this subset is fully optimized, just return the winner
+      return winner;
+    }
+
+    // if group is not fully explored. Its properties like cardinality
+    // would get changed after exploration. So it cannot return a valid LB
+    if (subset.getSet().getState() != CascadeRelSet.ExploreState.EXPLORED) {
+      return null;
+    }
+
+    RelOptCost lowerBound = null;
+    for (RelNode relNode : subset.getRels()) {
```

Comment:
If current group is not fully optimized and has no winner, there may be problems to get the lower bound here. Since the current group's optimization may affact the input groups of the current group. E.g., the current group's optimization may push down some require traits and  maybe somewhere of the inputs subtrees may enhance there's costs. So if we get the lower bound here, it may be larger than the actual lower bound which may miss optimal plan